azkaban-common/src/main/java/azkaban/jobExecutor/ProcessJob.java

```diff
@@ -259,7 +259,6 @@ public void run() throws Exception {
         }
         throw new RuntimeException(e);
       } finally {
-        this.process = null;
```

Comment:
TL;DR This is beyond the scope of the PR. Mostly my thoughts regarding this class.

I think we need a better way of managing the status of a process here. Currently what we have is a `boolean` called `killed` but effectively what we need is some level of granularity and a way to express it - similar to `status`.
There are certain benefits of maintaining a proper `status` for a process like `NOT_STARTED`, `STARTED`, `COMPLETE` or `KILLED` or `FAILED`. The last 3 being terminal nodes; think of this like a state machine. Having a simple mapping between the job status to the process status may make things much easier. Isn't it? Because a job will fail if a process fails, etc.
Using the process as `null` or not `null` as signals may not be the most intuitive way of doing it. I think this class is more like a `ProcessLifeCycleManager` which creates, runs and manages the lifecycle of the process. Currently, it subclasses from `Job` which is weird. Maybe the `Job` should have `ProcessLifeCycleManager` in it since every job runs in its own process. We should probably think of how we can restructure this code to make it simpler and more flexible.